bgzf.c

```diff
@@ -1922,10 +1935,24 @@ int bgzf_check_EOF(BGZF *fp) {
 
     if (fp->mt) {
         pthread_mutex_lock(&fp->mt->command_m);
+        // fp->mt->command state transitions should be:
+        // NONE -> HAS_EOF -> HAS_EOF_DONE -> NONE
+        // (HAS_EOF -> HAS_EOF_DONE happens in bgzf_mt_reader thread)
         fp->mt->command = HAS_EOF;
         pthread_cond_signal(&fp->mt->command_c);
         hts_tpool_wake_dispatch(fp->mt->out_queue);
-        pthread_cond_wait(&fp->mt->command_c, &fp->mt->command_m);
+        do {
+            pthread_cond_wait(&fp->mt->command_c, &fp->mt->command_m);
+            switch (fp->mt->command) {
+            case HAS_EOF_DONE: break;
+            case HAS_EOF:
+                pthread_cond_signal(&fp->mt->command_c);
```

Comment:
Maybe add some comments, such as "we caught the signal intended for bgzf_mt_reader, so send it again".

Also I wonder if the logic of *received in error, resend signal* should be in mt_reader instead.  Wouldn't this avoid the need for additional states in the state machine?  If mt_reader gets a NONE command then it knows it was its response and it can be ignored (resent).  That would keep the code receiving replies (main thread) simple.

Reply:
The same logic is [here](https://github.com/samtools/htslib/blob/5ebfddd/bgzf.c#L1567) in bgzf_mt_reader.  It can't signal on NONE because if it did it would get into an endless loop of signalling itself.  Signalling on `SEEK_DONE / HAS_EOF_DONE` it should get out of the loop as the other thread will at some point be scheduled and reset the command to `NONE`.

The logic here should also deal with the possibility of spurious wake-up, if an old signal was still pending for some reason.  Without this it's possible that the value of `fp->mt->eof` could be read before `bgzf_mt_reader()` has had a chance to change it.

Reply:
Comments also added.

Reply:
Not an endless loop - it's only endless if no one else is also trying to listen, which is an error in itself.  It's a loop of trying to not win the race, albeit perhaps not very efficient. :)  That said, if you win it once maybe there is some condition that makes you repeatedly win so you'd need some mitigation for that which becomes more complicated and possibly fragile.  Ie point taken.

Two communication channels (sender and receiver) would be cleanest, but it's also a more invasive change and more unknowns to worry about so this is sufficient.

Reply:
When `bgzf_mt_reader()` reaches EOF on a long file, it's likely that `mt->command` will be `NONE` and there will be no other listeners.  If it got past `pthread_cond_wait()` due to a pending signal and it signalled on NONE itself then it would start looping until the reader thread got around to signalling `CLOSE`.